### PR TITLE
Add status manager for object tree view.

### PR DESCRIPTION
**Pull request by sanebay**

### CODE REVIEW

File: include/sisl/status_mgr/status_mgr.hpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+namespace sisl {
+
+struct status_request {
+    nlohmann::json json;
+    bool do_recurse{false};
+    int verbose_level = 0;
+    std::vector< std::string > id_vec;
+};
+
+struct status_response {
```
Comment (szmyd) [inline include/sisl/status_mgr/status_mgr.hpp:37]: Is this meant to be extended beyond? Not sure why this isn't an alias.

File: include/sisl/status_mgr/status_mgr.hpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+using get_callback_type = std::function< status_response(const status_request&) >;
+class status_object;
+using status_object_ptr = std::shared_ptr< status_object >;
+
+// Keeps a heirarchy of modules/subsystems which register their callbacks to be called
+// whenever a get status is called from the root or directly.
+class status_object {
+    friend class status_manager;
```
Comment (szmyd) [inline include/sisl/status_mgr/status_mgr.hpp:48]: Is this needed? I don't see any access to `m_...` in status_manager.

File: include/sisl/status_mgr/status_mgr.hpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+public:
+    status_object(const std::string& name, const std::string& type, get_callback_type cb) :
+            m_name(name), m_type(type), m_get_status_cb(std::move(cb)) {}
```
Comment (szmyd) [inline include/sisl/status_mgr/status_mgr.hpp:52]: `cb` is not taken as an rvalue reference so this is just moving the copy. I think taking callbacks by copy is fine right? It's relatively cheap and happens only at startup.

File: include/sisl/status_mgr/status_mgr.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    static status_object_ptr create(const std::string& name, const std::string& type, get_callback_type cb) {
```
Comment (szmyd) [inline include/sisl/status_mgr/status_mgr.hpp:54]: Same as above here.

File: include/sisl/status_mgr/status_mgr.hpp (hunk 3 above)
Comment (sanebay) [reply include/sisl/status_mgr/status_mgr.hpp:52]: Yeah right

File: src/status_mgr/status_mgr.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+void status_manager::register_module(status_object_ptr module_obj) {
+    std::unique_lock lock{m_mtx};
+    m_module_object[module_obj->name()] = module_obj;
+    LOGINFO("Registered module {}", module_obj->name());
+}
+
+status_response status_manager::get_modules(const status_request& request) {
+    status_response response;
+    for (auto [module_name, status_obj] : m_module_object) {
```
[the remark this answers is not in the document]
Comment (sanebay) [reply src/status_mgr/status_mgr.cpp:31]: Ack.

File: src/status_mgr/status_mgr.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+status_response status_manager::get_module_status(const std::string& module_name, const status_request& request) {
+    auto iter = m_module_object.find(module_name);
+    if (iter == m_module_object.end()) { return status_error("Module not found"); }
+
+    status_object_ptr obj = iter->second;
+    status_response response;
+    if (!request.do_recurse) {
+        if (request.id_vec.empty()) { return status_error("No identifier found"); }
+
+        for (uint32_t ii = 1; ii < request.id_vec.size(); ii++) {
```
[the remark this answers is not in the document]
Comment (sanebay) [reply src/status_mgr/status_mgr.cpp:52]: The first member is found in the map and rest are found in its obj->children vec.

File: src/status_mgr/status_mgr.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    // Return all modules if no identifier given.
+    if (request.id_vec.empty()) {
+        return get_modules(request);
+    }
+
+    return get_module_status(request.id_vec[0] /* module_name */, request);
+}
+
+status_response status_manager::status_error(std::string error_str) {
```
[the remark this answers is not in the document]
Comment (sanebay) [reply src/status_mgr/status_mgr.cpp:73]: Ack

File: src/status_mgr/status_mgr.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    status_object_ptr obj = iter->second;
+    status_response response;
+    if (!request.do_recurse) {
+        if (request.id_vec.empty()) { return status_error("No identifier found"); }
+
+        for (uint32_t ii = 1; ii < request.id_vec.size(); ii++) {
+            obj = obj->get_child(request.id_vec[ii]);
+            if (obj == nullptr) { return status_error("Invalid identifier found"); }
+        }
+        return obj->get_status(request);
```
[the remark this answers is not in the document]
Comment (sanebay) [reply src/status_mgr/status_mgr.cpp:56]: We changed this.

**Pull request closed without merging**